src/charm.py

```diff
@@ -633,7 +669,66 @@ def _get_backup_key(self) -> str:
         timestamp = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
         return f"{BACKUP_KEY_PREFIX}-{self.model.name}-{timestamp}"
 
-    def _get_ca_cert_location_in_charm(self) -> str:
+    def _on_new_access_certificate_joined(self, event: RelationJoinedEvent):
+        if not self._ingress_address:
+            raise RuntimeError("Tried to process relation join before peer relation.")
+        if not (private_key := self._get_private_key_from_workload().encode()):
+            private_key = generate_private_key()
+        new_csr = generate_csr(private_key=private_key, subject=self._ingress_address)
+        self._push_csr_to_workload(new_csr.decode())
+        self._push_unit_pk_to_workload(private_key=private_key.decode())
+        self.tls_certificate_access.request_certificate_creation(
+            certificate_signing_request=new_csr
+        )
+        logger.info(
+            "New certificate requested from %s for unit %s",
+            event.relation.name,
+            self.unit.name,
+        )
+
+    def _on_new_access_certificate_available(self, event: CertificateAvailableEvent):
+        private_key = self._get_private_key_from_workload()
+        csr = self._get_csr_from_workload()
+        if not private_key or csr != event.certificate_signing_request:
+            return
+        self._push_unit_certificate_to_workload(certificate=event.certificate)
+        self._push_ca_certificate_to_workload(certificate=event.ca)
+        self._container.send_signal(SIGHUP, self._container_name)
+
+    def _on_access_certificate_expiring(self, event: CertificateExpiringEvent):
+        private_key = self._get_private_key_from_workload()
+        old_csr = self._get_csr_from_workload()
+        old_cert = self._get_cert_from_workload()
+        if event.certificate != old_cert:
+            return
+        if not self._ingress_address:
+            raise RuntimeError("Peer relation somehow unavailable. Something is very wrong")
```

Comment:
I'm not sure we want the whole charm to crash because of this certificate problem, I'd recommend logging an error instead.

Reply:
Ah, I just assumed we always raised because we always raise it right now but I guess this case is different. How should I handle a nonexistent ingress address then since we need it here other than deferring? It's pretty much impossible for it to happen here, but the type hints are complaining if I don't deal with it.

Reply:
This is why I'm pointing to moving logic into `_configure`. Multiple events will cause this code to run and for each of them we will look at whether there's an action to take and take it if necessary.

Reply:
See comment above